### ISSUE TITLE
Why always assume the model going have id as primary key? why can not we dynamically get PK field from model?

### ISSUE THREAD

**Issue by BabuDip** [django-import-export/django-import-export#1818]
I have been using this library for some time now and it is fantastic how i works. It always bugged me why we choose to hardcode id as primary key instead of dynamically fetching it from the model schema?

In recent release, I realise that you are trying to provide use with choices to select fields, which is nice to have, however, it also expect primary key name as id. 

**I have a field name code which is set as primary key, and it breaks while exporting:**
AttributeError at /ccos-admin/ips_cca/departments/export/
'Departments' object has no attribute 'id'
```
/py/lib/python3.12/site-packages/import_export/admin.py, line 736, in export_action
            formats,
            self.get_export_resource_classes(request),
            data=request.POST or None,
        )
        form.fields["export_items"] = MultipleChoiceField(
            widget=MultipleHiddenInput,
            required=False,
            choices=[(o.id, o.id) for o in self.model.objects.all()],
                           ^^^^ …
        )
        if form.is_valid():
            file_format = formats[int(form.cleaned_data["format"])]()
            if "export_items" in form.changed_data:
                # this request has arisen from an Admin UI action
```
**Below is what overwrite and it works now. Sorry, I have not contributed and not familiar with pull request yet.**
```python
    def export_action(self, request):
        """
        Override export action to use primary key with any name.
        """
        if not self.has_export_permission(request):
            raise PermissionDenied
        # get pk field from the model
        pk_field = self.model._meta.pk.name

        formats = self.get_export_formats()
        form_class = self.get_export_form_class()
        form = form_class(
            formats,
            self.get_export_resource_classes(request),
            data=request.POST or None,
        )

        # Change the generation of choices to use the custom primary key.
        form.fields["export_items"] = forms.MultipleChoiceField(
            widget=forms.MultipleHiddenInput(),
            required=False,
            choices=[
                (getattr(o, pk_field), getattr(o, pk_field))
                for o in self.model.objects.all()
            ],
        )

        if form.is_valid():
            file_format = formats[int(form.cleaned_data["format"])]()
            queryset = (
                self.model.objects.filter(
                    **{pk_field + "__in": form.cleaned_data["export_items"]}
                )
                if "export_items" in form.changed_data
                else self.get_export_queryset(request)
            )

            try:
                export_data = self.get_export_data(
                    file_format,
                    request,
                    queryset,
                    encoding=self.to_encoding,
                    export_form=form,
                )
                content_type = file_format.get_content_type()
                response = HttpResponse(export_data, content_type=content_type)
                response["Content-Disposition"] = 'attachment; filename="%s"' % (
                    self.get_export_filename(request, queryset, file_format),
                )
                return response
            except Exception as e:
                messages.error(request, str(e))

        context = {"form": form, "opts": self.model._meta, "title": _("Export")}
        return TemplateResponse(request, self.export_template_name, context)
```        
        
**Also for the import, I have defined import_id_fields in meta, and still gets error which was working in previous version.
        Errors**
```
The following fields are declared in 'import_id_fields' but are not present in the resource fields: id
Traceback (most recent call last):
File "/py/lib/python3.12/site-packages/import_export/resources.py", line 871, in import_data_inner
self._check_import_id_fields(dataset.headers)
File "/py/lib/python3.12/site-packages/import_export/resources.py", line 1129, in _check_import_id_fields
raise exceptions.FieldError(
import_export.exceptions.FieldError: The following fields are declared in 'import_id_fields' but are not present in the resource fields: id
```

**Please let me know if you need any further information.**

**Comment by matthewhegarty** [comment]
Thanks for raising this.  Can you try with [this PR](https://github.com/django-import-export/django-import-export/pull/1802) and see if it solves your issue?

**Comment by BabuDip** [comment]
@matthewhegarty 
Thanks for the swift response. I have tested it and it works for export with custom pk.
I am still having import issues, sorry I should have split it into another issue.

**Comment by matthewhegarty** [comment]
Please can you test if [this PR](https://github.com/django-import-export/django-import-export/pull/1813) fixes your import issue?

**Comment by BabuDip** [comment]
@matthewhegarty I have tested [this PR](https://github.com/django-import-export/django-import-export/pull/1813) and it does not work in my case, Trying to import data with custom PK field.

**Comment by matthewhegarty** [comment]
Do you have `import_id_fields = "id"` defined in your resource?  Is it possible you could create a new issue with steps to reproduce?

**Comment by BabuDip** [comment]
Below is the steps to reproduce:

**requirements.txt**
`
django-import-export==4.0.1
django==5.0.6
openpyxl==3.1.2
`

**models.py**
`
from django.db import models

class Division(models.Model):
    id = models.AutoField(primary_key=True)
    name = models.CharField(max_length=100)
    description = models.TextField()
    created_at = models.DateTimeField(auto_now_add=True)
    updated_at = models.DateTimeField(auto_now=True)

    def __str__(self):
        return self.name

    class Meta:
        verbose_name_plural = "Divisions"


class Site(models.Model):
    division = models.ForeignKey(Division, on_delete=models.CASCADE)
    code = models.CharField(max_length=10, primary_key=True)
    name = models.CharField(max_length=100)
    description = models.TextField()
    created_at = models.DateTimeField(auto_now_add=True)
    updated_at = models.DateTimeField(auto_now=True)

    def __str__(self):
        return self.name

    class Meta:
        verbose_name_plural = "Sites"
`
**admin.py**
`
from django.contrib import admin
from import_export import resources
from import_export.admin import ImportExportMixin

from .models import Division, Site


class DivisionResource(resources.ModelResource):
    class Meta:
        model = Division
        fields = ("id", "name", "description", "created_at", "updated_at")


@admin.register(Division)
class DivisionAdmin(ImportExportMixin, admin.ModelAdmin):
    resource_class = DivisionResource
    list_display = ("id", "name", "description", "created_at", "updated_at")
    search_fields = ("name", "description")


class SiteResource(resources.ModelResource):
    class Meta:
        model = Site
        fields = ("division", "code", "name", "description", "created_at", "updated_at")
        import_id_fields = ("code",)


@admin.register(Site)
class SiteAdmin(ImportExportMixin, admin.ModelAdmin):
    resource_class = SiteResource
    list_display = (
        "division",
        "code",
        "name",
        "description",
        "created_at",
        "updated_at",
    )
    search_fields = ("code", "name", "description")
`
Everything works fine, except importing site (which has pk field as code)
![image](https://github.com/django-import-export/django-import-export/assets/71445383/ddcb9a25-a22f-4808-9b59-217b3675f527)

**Comment by matthewhegarty** [comment]
Can you add this to `SiteAdmin` and try again to see a full stack trace:

```
import_error_display = ("message", "row", "traceback")
```

Can you post an example row of your import file.

**Comment by BabuDip** [comment]
Hi @matthewhegarty,

Sure. Here is the walkthrough:
Once the server (with above config) is ready I added a site with below details:
![image](https://github.com/django-import-export/django-import-export/assets/71445383/ee958aec-03e7-4d63-8f46-433c2648e685)

To test the basic workflow, just exported the data as is:
![image](https://github.com/django-import-export/django-import-export/assets/71445383/d9411ef9-bd51-494a-9ed5-4fee51b8cb63)


which looks something like below:
![image](https://github.com/django-import-export/django-import-export/assets/71445383/94f62d7f-798d-45f8-9c62-f732be57e0c8)


and when I try to import the same file, it prints the below error (with import_error_display = ("message", "row", "traceback")):
![image](https://github.com/django-import-export/django-import-export/assets/71445383/df405c85-7021-4504-9e51-257e01eff267)

Please let me know if you need any further information.

Thank you.

**Comment by matthewhegarty** [comment]
I can't see what is causing this.  You have the following declaration:

```python
class SiteResource(resources.ModelResource):
    class Meta:
        model = Site
        fields = ("division", "code", "name", "description", "created_at", "updated_at")
        import_id_fields = ("code",)
```

Which clearly defines the `import_id_fields`, meaning that `id` should not be used.  Are you definitely using v4.0.1?  If you set a breakpoint [here](https://github.com/django-import-export/django-import-export/blob/6eb2d21fbefe97e8e85e94bc22d655a8172f15d4/import_export/resources.py#L192) can you verify that 'code' is being returned as the `import_id_field`?

Try declaring the following in `SiteAdmin` to rule out any issues there.

```
resource_classes = [SiteResource]
```

Can you step through with a debugger and isolate what is happening?

**Comment by BabuDip** [comment]
Hi @matthewhegarty, with version 4.0.1, it gets below:
self._meta.import_id_fields = ['id']

however, with recent version 4.0.3, it is now correct:
self._meta.import_id_fields = ('code',)

Thank you. All resolved now.
